crates/node/src/storage/rocksdb.rs

```diff
@@ -688,12 +734,148 @@ impl RocksDB {
+        let bytes = format!("chunks:{val}");
+        buf.write_all(bytes.as_bytes())?;
+        buf.flush()?;
+        Ok(())
+    }
+
+    /// Remove snapshots older than the latest
+    pub fn cleanup(
+        latest_height: BlockHeight,
+        base_dir: &Path,
+    ) -> std::io::Result<()> {
+        let toml = OsStr::new("toml");
+        for entry in std::fs::read_dir(base_dir)? {
+            let entry = entry?;
+            if entry.path().is_file() && Some(toml) == entry.path().extension()
+            {
+                if let Some(name) = entry.path().file_name() {
+                    let Some(height) = name
+                        .to_string_lossy()
+                        .strip_prefix("snapshot_")
+                        .and_then(|n| n.strip_suffix(".toml"))
+                        .and_then(|h| BlockHeight::from_str(h).ok())
+                    else {
+                        continue;
+                    };
+                    if height < latest_height {
+                        _ = std::fs::remove_file(entry.path());
+                    }
+                };
+            }
+        }
+        Ok(())
+    }
+
+    /// Create a path to save a snapshot at a specific block height.
+    pub fn path(height: BlockHeight, mut base_dir: PathBuf) -> PathBuf {
+        base_dir.push(format!("snapshot_{}.toml", height));
+        base_dir
+    }
+}
+
+/// A chunk of a snapshot. Includes the last line number in the file
+/// for this chunk and a hash of the chunk contents.
+#[derive(
+    Debug, Clone, Default, PartialEq, Eq, BorshSerialize, BorshDeserialize,
+)]
+struct Chunk {
+    boundary: u64,
+    hash: Hash,
+}
+
+/// Builds a set of chunks from a stream of lines to be
+/// written to a file.
+#[derive(Debug, Clone)]
+struct Chunker {
+    chunks: Vec<Chunk>,
+    max_size: usize,
+    current_boundary: u64,
+    current_size: usize,
+    hasher: Sha256,
+}
+impl Chunker {
+    fn new(max_size: usize) -> Self {
+        Self {
+            chunks: vec![],
+            max_size,
+            current_boundary: 0,
+            current_size: 0,
+            hasher: Sha256::default(),
+        }
+    }
+
+    fn add_line(&mut self, line: &str) {
+        if checked!(self.current_size + line.as_bytes().len()).unwrap()
+            > self.max_size
+            && self.current_boundary != 0
+        {
+            let mut hasher = Sha256::default();
+            std::mem::swap(&mut hasher, &mut self.hasher);
+            let hash: [u8; 32] = hasher.finalize().into();
+            self.chunks.push(Chunk {
+                boundary: self.current_boundary,
+                hash: Hash(hash),
+            });
+            self.current_size = 0;
+        }
+
+        self.current_size =
+            checked!(self.current_size + line.as_bytes().len()).unwrap();
+        self.hasher.update(line.as_bytes());
+        self.current_boundary = checked!(self.current_boundary + 1).unwrap();
```

Comment:
if you're based on `v0.39.0` we already have `checked!(x += y)` btw
```suggestion
        checked!(self.current_boundary += 1).unwrap();
```